### PR TITLE
SI-8818 FreshName extractor forgives suffix

### PR DESCRIPTION
**Pull request by som-snytt**
The test is corrected (inverted) and the extractor is made
more succinct.  Succinctness isn't enforced by the test,
but I checked it manually.

### CODE REVIEW

File: src/reflect/scala/reflect/internal/FreshNames.scala [hunk 1]
```diff
@@ -23,17 +24,20 @@ trait FreshNames { self: Names with StdNames =>
...
+
+    // name should start with creatorPrefix and end with number
+    val freshlyNamed = {
+      val pre = if (!creatorPrefix.isEmpty) Regex quote creatorPrefix else ""
+      s"""$pre(.*?)\\d*""".r
     }
+
+    def unapply(name: Name): Option[String] =
+      name.toString match {
+        case freshlyNamed(prefix) => Some(prefix)
```
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/FreshNames.scala:39]: Friday musing: wouldn't it be nice to have a `toOption` on `Seq` that gives you `Some(headIHaveNoTail)` or `None`?

```
def unapply(name: Name): Option[String] = freshlyNamed.unapplySeq(name).flatMap(_.toOption)
```

**Comment by adriaanm** [comment]
LGTM

File: src/reflect/scala/reflect/internal/FreshNames.scala (hunk 1 above)
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/FreshNames.scala:39]: `tailsYouLose`? as they say in coin flippage. `xs.headOption filter (_ => xs.tail.isEmpty)`.

File: test/junit/scala/tools/nsc/symtab/FreshNameExtractorTest.scala [hunk 2]
```diff
@@ -32,16 +32,16 @@ class FreshNameExtractorTest {
     val Creator = new FreshNameCreator(prefixes.head)
     val Extractor = new FreshNameExtractor(prefixes.tail.head)
     assertThrows[MatchError] {
-      val Extractor(_) = TermName(Creator.newName("foo"))
+      TermName(Creator.newName("foo")) match { case Extractor(_) => }
```
Comment (som-snytt) [inline test/junit/scala/tools/nsc/symtab/FreshNameExtractorTest.scala:35]: Sorry I made this less cool.

**Comment by som-snytt** [comment]
@adriaanm :interrobang:

Hello from the front of @adriaanm 's :mailbox:

**Comment by adriaanm** [comment]
:fire_engine:

where's the :fire:?

**Comment by adriaanm** [comment]
Or is is :cat: in a :christmas_tree:?

**Comment by adriaanm** [comment]
See also #4304 for a more recent version.

**Comment by som-snytt** [comment]
OK, thx @adriaanm .  I was freebasing yesterday but didn't think :tested: (and apparently that's not an emoji) stuff needed it. Still working on my emoji fu. Apparently fu is not the emoji for emoji fu.

**Comment by lrytz** [comment]
here's some emoji fu: http://www.theguardian.com/us-news/ng-interactive/2015/jan/20/-sp-state-of-the-union-2015-address-obama-emoji

**Pull request closed without merging**